### DISCUSSION TITLE
Prediksi Hasil Clustering ke Data Test

### DISCUSSION THREAD

**Discussion by joejnt** [teamalgoritma/community#869] [Q&A] [answered]
Dear rekan2 algoritma,

Mohon bantuannya,
Saya mendapatkan permasalahan pada saat melakukan clustering menggunakan K-Means, dimana memory used di R nya sudah mencapai 12 GiB sehingga pada saat saya melakukan proses clustering hasilnya gagal. Klo saya tanyakan ke GPT penyebabnya adalah karena proses PCA yang menggunakan kapasitas storage yang cukup besar. Kemudian saya melakukan beberapa langkah, yaitu data hasil PCA saya simpan dalam format rds kemudian syntax untuk proses PCA tidak saya jalankan. Hasilnya penggunaan storage berkurang namun proses clustering masih tetap tidak bisa karena permasalahan storage. Kemudian saya melakukan proses split data dengan proporsi 0.7, rencananya saya akan melakukan proses clustering pada data train kemudian data test saya anggap sebagai data baru. namun pada saat saya melakukan prediksi hasilnya error. Mohon sarannya, adapun syntax yang saya gunakan adalah sebagai berikut:

1. Proses PCA:
data_scaled <- as.data.frame(scale(dfCL %>% select(c(freq, volume, high, low))))
df_pca <- PCA(data_scaled, quali.sup = 4, graph = F)
df_pca$eig
saveRDS(as.data.frame(df_pca$ind$coord[,c(1:2)]), "Data Output/d_pca.rds")

2. Split dataset
d_pca <- readRDS("Data Output/d_pca.rds")
df_pca <- cbind(dfCL %>% select(Store_Name), d_pca)

split <- sample.split(df_pca, SplitRatio = 0.7)
train <- subset(df_pca, split == TRUE)
test <- subset(df_pca, split == FALSE)

3. Clustering
set.seed(100)
eu_pca_km <- eclust(train %>% select(-Store_Name),
                    stand = TRUE,
                    FUNcluster = "kmeans",
                    k = 3, 
                    graph = F,
                    hc_metric = "euclidean")

4. Prediksi
predicted_clusters_eu <- predict(eu_pca_km$cluster, newdata = test %>% select_if(is.numeric))

**Comment by kinantytasyao** [comment]
Dear @joejnt 

Terima kasih telah bertanya di github community Algoritma. Apakah boleh dilampirkan screenshot dari error prediksi yang dilampirkan agar kami dapat dengan efektif membantu menyelesaikan errornya?

**Reply by joejnt** [reply]
Dear @kinantytasyao 

Untuk screenshotnya saya ngga punya, hanya beritanya saja:
Error in UseMethod("predict") : 
  no applicable method for 'predict' applied to an object of class "c('kmeans', 'eclust')"

Apakah ini cukup ya ?

Best regards,
JT

**Reply by saskia-dwi-ulfah** [reply, marked as answer]
Dear @joejnt 

Terima kasih telah bertanya di Github Comminity dan atas kesediaannya untuk menunggu. 

Berdasarkan pesan error yang Anda lampirkan, method `predict()` tidak dapat digunakan secara langsung untuk model clustering yang sudah dibuat sebelumnya. Ada beberapa alternatif yang bisa dieksplor disini:

* Alternatif 1 dan 2: Berdasarkan [referensi ini](https://stats.stackexchange.com/questions/78322/is-there-a-function-in-r-that-takes-the-centers-of-clusters-that-were-found-and), Anda dapat membuat fungsi custom di mana fungsi ini akan menghitung jarak dari data baru ke cluster yang paling dekat. Alternatif lainnya, Anda bisa menggunakan fungsi `predict()` dari package [flexclust](https://cran.r-project.org/web/packages/flexclust/index.html).

Semoga jawaban ini dapat membantu.

Best Regards,
Team Algoritma

**Reply by joejnt** [reply]
@saskia-dwi-ulfah 

Terima kasih untuk inside nya.

Best regards,
JT
